concurrent/src/main/java/io/airlift/concurrent/DynamicSizeBoundQueue.java

```diff
@@ -0,0 +1,337 @@
+     * on numeric overflow. This is slightly less efficient than the normal getAndAdd (which often has intrinsic
+     * support). If this ever becomes a performance bottleneck, it is possible to use the original getAndAdd if
+     * the caller can guarantee no risk of numeric overflow.
+     */
+    private static long getAndAddChecked(AtomicLong atomicLong, long delta)
+    {
+        return atomicLong.getAndAccumulate(delta, Math::addExact);
+    }
+
+    public boolean offer(T element, long timeout, TimeUnit unit)
+            throws InterruptedException
+    {
+        long elementSize = elementSizeFunction.applyAsLong(element);
+        long remainingTimeoutNs = unit.toNanos(timeout);
+        while (!offer(element, elementSize)) {
+            ListenableFuture<Void> future = getOrCreateFuture(dequeueFuture);
+            // Check again in case we already missed the relevant dequeue event
+            if (offer(element, elementSize)) {
+                break;
+            }
+            long startTimeNs = ticker.read();
+            if (remainingTimeoutNs <= 0 || !awaitDequeueFuture(future, remainingTimeoutNs, NANOSECONDS)) {
+                // Timed out
+                return false;
+            }
+            remainingTimeoutNs -= ticker.read() - startTimeNs;
+        }
+        return true;
+    }
+
+    public void put(T element)
+            throws InterruptedException
+    {
+        long elementSize = elementSizeFunction.applyAsLong(element);
+        while (!offer(element, elementSize)) {
+            ListenableFuture<Void> future = getOrCreateFuture(dequeueFuture);
+            // Check again in case we already missed the relevant dequeue event
+            if (offer(element, elementSize)) {
+                break;
+            }
+            awaitDequeueFuture(future);
+        }
+    }
+
+    /**
+     * Enqueue the element if there is space, otherwise returns a ListenableFuture that will complete
+     * when space becomes available for the element. If a future is returned, the element was not inserted.
+     */
+    public Optional<ListenableFuture<Void>> offerWithBackoff(T element)
+    {
+        long elementSize = elementSizeFunction.applyAsLong(element);
+        if (offer(element, elementSize)) {
+            return Optional.empty();
+        }
+        ListenableFuture<Void> future = getOrCreateFuture(dequeueFuture);
+        // Check again in case we already missed the relevant dequeue event
+        if (offer(element, elementSize)) {
+            return Optional.empty();
+        }
+        return Optional.of(Futures.nonCancellationPropagating(future));
+    }
+
+    /**
+     * Insert without regard to the max size (potentially exceeding the max limit)
+     */
+    public void forcePut(T element)
+    {
+        long elementSize = elementSizeFunction.applyAsLong(element);
+        checkArgument(elementSize > 0, "element size must be positive");
+        getAndAddChecked(size, elementSize);
```

Comment:
This will throw on overflow, which at the very least should be documented.  Since this is typically used to add poison pill, the user should be encouraged (forced?) to only use this with elements that have a zero size

Reply:
throw IllegalStateException on arithmetic overflow